File: Part3/emergence_sim.f90
program emergence_sim_fortran_dynamic
    use, intrinsic :: iso_fortran_env, only: real64, int32
    implicit none

    integer :: DIM, N_AGENTS
    real(real64), parameter :: EPSILON_METRIC = 1e-5
    real(real64), parameter :: EPSILON_MANIFOLD = 1e-4
    real(real64), parameter :: DT = 0.1_real64
    real(real64), parameter :: PI = 3.1415926535897932_real64

    type :: Agent
        integer :: idx
        real(real64), allocatable :: x(:)
        real(real64), allocatable :: K(:, :)
        real(real64) :: A_val, hunger, reward
        real(real64) :: C_vec(3)
        character(len=12) :: mode
        real(real64) :: h_threshold, t_threshold, e_power
        real(real64), allocatable :: affinities(:)
    end type Agent

    type :: World
        type(Agent), allocatable :: agents(:)
        real(real64), allocatable :: resource_pos(:)
        real(real64) :: l_base
    end type World

    type(World) :: w
    integer, parameter :: LOG_UNIT = 15
    character(len=256) :: log_dir = "/private/test/PKGF_Intelligence_Emergence/Part3/logs/fortran"

    call execute_part3_comparison()

contains

    subroutine execute_part3_comparison()
        integer :: n_list(3) = [4, 8, 16]
        integer :: i
        character(len=256) :: log_path
        
        do i = 1, 3
            N_AGENTS = n_list(i)
            DIM = n_list(i)
            write(*, '(A, I2, A, I2)') "--- Executing Case: N=", N_AGENTS, ", DIM=", DIM
            
            log_path = trim(log_dir) // "/case_N" // char(48 + N_AGENTS/10) // char(48 + mod(N_AGENTS,10)) // ".txt"
            open(unit=LOG_UNIT, file=trim(log_path), status='replace', action='write')
            call run_single_case()
            close(LOG_UNIT)
        end do
    end subroutine execute_part3_comparison

    subroutine run_single_case()
        integer :: s, j, a_count, n_count, s_count
        real(real64) :: avg_r, avg_t
        
        call init_world_dynamic(N_AGENTS, DIM)
        write(LOG_UNIT, '(A, I2, A, I2)') "--- Case: N=", N_AGENTS, ", DIM=", DIM
        write(LOG_UNIT, '(A)') "Step  | Modes (A/N/S)   | Avg_Rew    | Avg_Ten"
        
        do s = 0, 200
            a_count = 0; n_count = 0; s_count = 0
            avg_r = 0.0; avg_t = 0.0
            do j = 1, N_AGENTS
                if (trim(w%agents(j)%mode) == "Aggressive") a_count = a_count + 1
                if (trim(w%agents(j)%mode) == "Neutral") n_count = n_count + 1
                if (trim(w%agents(j)%mode) == "Submissive") s_count = s_count + 1
                avg_r = avg_r + w%agents(j)%reward
                avg_t = avg_t + w%agents(j)%C_vec(3)
            end do
            avg_r = avg_r / real(N_AGENTS, real64)
            avg_t = avg_t / real(N_AGENTS, real64)
            
            if (mod(s, 40) == 0) then
                write(LOG_UNIT, '(I5, " | ", I2, "/", I2, "/", I2, " | ", F10.4, " | ", F10.4)') &
                    s, a_count, n_count, s_count, avg_r, avg_t
                write(*, '(I5, " | ", I2, "/", I2, "/", I2, " | ", F10.4, " | ", F10.4)') &
                    s, a_count, n_count, s_count, avg_r, avg_t
            end if
            call world_step_dynamic()
        end do
    end subroutine run_single_case

    subroutine init_world_dynamic(n, d)
        integer, intent(in) :: n, d
        integer :: i, j, k
        real(real64) :: r_v, ratio, diff
        if (allocated(w%agents)) deallocate(w%agents)
        allocate(w%agents(n))
        if (allocated(w%resource_pos)) deallocate(w%resource_pos)
        allocate(w%resource_pos(d))
        w%resource_pos = 0.0
        w%l_base = 0.4
        
        do i = 1, n
            w%agents(i)%idx = i
            allocate(w%agents(i)%x(d))
            allocate(w%agents(i)%K(d, d))
            allocate(w%agents(i)%affinities(n))
            
            do k = 1, d
                call random_number(r_v)
                w%agents(i)%x(k) = (r_v - 0.5_real64) * 2.0_real64
            end do
            
            w%agents(i)%K = 0.0
            do j = 1, d; w%agents(i)%K(j,j) = 1.0; end do
            
            w%agents(i)%A_val = 0.0; w%agents(i)%hunger = 0.0; w%agents(i)%reward = 0.0
            w%agents(i)%C_vec = [1.0, 1.0, 0.0]; w%agents(i)%mode = "Neutral"
            
            ratio = real(i-1, real64) / real(n-1, real64)
            if (ratio < 0.2) then
                w%agents(i)%h_threshold = 0.2; w%agents(i)%t_threshold = 5.0; w%agents(i)%e_power = 0.01
            else if (ratio < 0.5) then
                w%agents(i)%h_threshold = 0.6; w%agents(i)%t_threshold = 1.2; w%agents(i)%e_power = 0.8
            else
                w%agents(i)%h_threshold = 1.2; w%agents(i)%t_threshold = 0.6; w%agents(i)%e_power = 5.0
            end if
            
            do j = 1, n
                diff = min(real(abs(i-j), real64), real(n-abs(i-j), real64))
                w%agents(i)%affinities(j) = cos(2.0_real64 * PI * diff / real(n, real64))
            end do
        end do
    end subroutine init_world_dynamic

    subroutine world_step_dynamic()
        real(real64), allocatable :: v_list(:,:)
        integer :: i
        allocate(v_list(DIM, N_AGENTS))
        do i = 1, N_AGENTS; v_list(:,i) = get_v_dynamic(w%agents(i)); end do
        do i = 1, N_AGENTS; call decide_dynamic(w%agents(i)); end do
        do i = 1, N_AGENTS; call update_dynamic(w%agents(i), v_list(:,i)); end do
        deallocate(v_list)
    end subroutine world_step_dynamic

    function get_v_dynamic(a) result(v_out)
        type(Agent), intent(in) :: a
        real(real64) :: v_out(DIM), v_pkgf(DIM), v_des(DIM), v_eth(DIM), eta(DIM), r_v, d_w, e_s, dist_sq, aff
        real(real64) :: om_x(DIM), om_p(DIM), pp(DIM)
        integer :: i, o
        
        ! Co-differential Approx
        om_x = get_om_dynamic(a%x)
        v_pkgf = 0.0
        do i = 1, DIM
            pp = a%x; pp(i) = pp(i) + 1e-4_real64; om_p = get_om_dynamic(pp)
            v_pkgf = v_pkgf + (om_p - om_x) / 1e-4_real64
        end do
        
        d_w = 0.3_real64 + a%hunger * 3.0_real64
        v_des = (w%resource_pos - a%x) * d_w
        
        v_eth = 0.0
        do o = 1, N_AGENTS
            dist_sq = sum((a%x-w%agents(o)%x)**2)
            if (dist_sq < 1e-9) cycle
            aff = a%affinities(o)
            e_s = a%e_power * (1.5_real64 - aff)
            if (trim(a%mode) == "Aggressive") e_s = e_s * 0.01
            if (trim(a%mode) == "Submissive") e_s = e_s * 2.0
            
            v_eth = v_eth - e_s * (2.0_real64 / real(N_AGENTS, real64)) * &
                    (1.0_real64 + a%C_vec(3)) / (dist_sq + 0.05_real64) * (a%x-w%agents(o)%x)
            if (aff > 0.5) v_eth = v_eth + 0.2_real64 * aff / (sqrt(dist_sq) + 1.0_real64) * (w%agents(o)%x-a%x)
        end do
        
        do i = 1, DIM; call random_number(r_v); eta(i) = (r_v*0.02-0.01); end do
        v_out = v_pkgf + v_des + w%l_base * v_eth + eta
    end function get_v_dynamic

    function get_om_dynamic(pos) result(om)
        real(real64), intent(in) :: pos(DIM)
        real(real64) :: om(DIM), g(DIM, DIM)
        g = get_metric_dynamic(pos); om = matmul(g, (w%resource_pos - pos) * 0.4_real64)
    end function get_om_dynamic

    function get_metric_dynamic(x) result(g)
        real(real64), intent(in) :: x(DIM)
        real(real64) :: g(DIM, DIM), s_ctx, avg_s
        integer :: i, ctx_idx
        g = 0.0; ctx_idx = (DIM * 3) / 4 + 1
        avg_s = 1.0
        if (ctx_idx <= DIM) then
            s_ctx = 0.0
            do i = ctx_idx, DIM; s_ctx = s_ctx + (1.0_real64 + 0.5_real64 * tanh(x(i))); end do
            avg_s = s_ctx / real(DIM - ctx_idx + 1, real64)
        end if
        do i = 1, DIM
            if (i < ctx_idx) then; g(i,i) = avg_s; else; g(i,i) = 1.0; end if
        end do
    end function get_metric_dynamic

    subroutine decide_dynamic(a)
        type(Agent), intent(inout) :: a
        real(real64) :: stress, dist
        integer :: o
        stress = 0.0
        do o = 1, N_AGENTS
            if (o == a%idx) cycle
            dist = sqrt(sum((a%x-w%agents(o)%x)**2)) + 1e-9
            stress = stress - a%affinities(o) / (dist + 0.5_real64)
        end do
        a%C_vec(3) = max(0.0_real64, min(2.0_real64, a%C_vec(3) + stress * 0.02)) ! Temporal tension shift
        
        if (a%hunger > a%h_threshold) then; a%mode = "Aggressive"
        else if (a%C_vec(3) > a%t_threshold) then; a%mode = "Submissive"
        else; a%mode = "Neutral"; end if
        if (a%reward > 0.1) a%hunger = a%hunger * 0.5
    end subroutine decide_dynamic

    subroutine update_dynamic(a, v)
        type(Agent), intent(inout) :: a
        real(real64), intent(in) :: v(DIM)
        real(real64) :: d_p, d_c, ex, tau, r_v
        integer :: i
        d_p = sqrt(sum((a%x - w%resource_pos)**2))
        a%x = a%x + v * DT
        d_c = sqrt(sum((a%x - w%resource_pos)**2))
        
        call transport_K_dynamic(a%K, a%x - v*DT, v)
        
        ex = 1.2_real64 * max(0.0_real64, (d_c-d_p)/DT) + 0.4_real64 * sqrt(sum(v**2))
        a%A_val = max(0.0_real64, min(5.0_real64, a%A_val + (-0.6_real64*a%A_val+ex)*DT))
        do i = 1, DIM; call random_number(r_v); a%K(i,i) = a%K(i,i) + (r_v*0.01-0.005)*a%A_val; end do
        
        tau = 0.35_real64
        a%C_vec(3) = a%C_vec(3) + (-tau*a%C_vec(3) + tanh(a%A_val*1.3_real64)) * DT
        a%C_vec(3) = max(0.0_real64, min(2.0_real64, a%C_vec(3)))
        
        if (d_c < 0.4) then; a%reward = a%reward + 0.4/(1.0+a%A_val)
        else; a%hunger = a%hunger + 0.15*DT; end if
    end subroutine update_dynamic

    subroutine transport_K_dynamic(K, x, v)
        real(real64), intent(inout) :: K(:, :)
        real(real64), intent(in) :: x(:), v(:)
        real(real64), allocatable :: Gam(:, :, :), Om(:, :), H(:, :)
        integer :: i_t, j_t, k_t, n_t
        n_t = size(x)
        allocate(Gam(n_t, n_t, n_t), Om(n_t, n_t), H(n_t, n_t))
        call get_gam_dynamic(x, Gam)
        Om = 0.0
        do i_t = 1, n_t; do j_t = 1, n_t; do k_t = 1, n_t
            Om(i_t, j_t) = Om(i_t, j_t) + Gam(i_t, k_t, j_t) * v(k_t)
        end do; end do; end do
        H = exp_m_dynamic(Om * DT)
        K = matmul(matmul(H, K), mat_i_dynamic(H))
        deallocate(Gam, Om, H)
    end subroutine transport_K_dynamic

    subroutine get_gam_dynamic(x, G)
        real(real64), intent(in) :: x(:)
        real(real64), intent(out) :: G(:, :, :)
        real(real64), allocatable :: gi(:, :), pg(:, :, :), xp(:), xm(:)
        integer :: i_g, j_g, k_g, l_g, n_g
        n_g = size(x)
        allocate(gi(n_g, n_g), pg(n_g, n_g, n_g), xp(n_g), xm(n_g))
        gi = mat_i_dynamic(get_metric_dynamic(x))
        do k_g = 1, n_g
            xp = x; xp(k_g) = xp(k_g) + EPSILON_METRIC
            xm = x; xm(k_g) = xm(k_g) - EPSILON_METRIC
            pg(k_g, :, :) = (get_metric_dynamic(xp) - get_metric_dynamic(xm)) * (0.5_real64 / EPSILON_METRIC)
        end do
        G = 0.0
        do k_g = 1, n_g; do i_g = 1, n_g; do j_g = 1, n_g; do l_g = 1, n_g
            G(k_g, i_g, j_g) = G(k_g, i_g, j_g) + 0.5_real64 * gi(k_g, l_g) * (pg(i_g, j_g, l_g) + pg(j_g, i_g, l_g) - pg(l_g, i_g, j_g))
        end do; end do; end do; end do
        deallocate(gi, pg, xp, xm)
    end subroutine get_gam_dynamic

    function mat_i_dynamic(A) result(Ai)
        real(real64), intent(in) :: A(:, :)
        real(real64) :: Ai(size(A,1), size(A,1)), m(size(A,1), 2*size(A,1)), p_v, f
        integer :: i, k, n
        n = size(A, 1)
        Ai = 0.0; do i = 1, n; Ai(i, i) = 1.0; end do
        m(:, 1:n) = A; m(:, n+1:2*n) = Ai
        do i = 1, n
            p_v = m(i, i); m(i, :) = m(i, :) / p_v
            do k = 1, n; if (i /= k) then; f = m(k, i); m(k, :) = m(k, :) - f * m(i, :); end if; end do
        end do
        Ai = m(:, n+1:2*n)
    end function mat_i_dynamic

    function exp_m_dynamic(A) result(E)
        real(real64), intent(in) :: A(:, :)
        real(real64) :: E(size(A,1), size(A,1)), As(size(A,1), size(A,1)), Ap(size(A,1), size(A,1))
        real(real64) :: P(size(A,1), size(A,1)), Q(size(A,1), size(A,1)), n_inf
        real(real64) :: c(7) = [1.0, 0.5, 0.1, 0.01111111111111111, 0.0007575757575757576, 3.0303030303030303e-05, 5.050505050505051e-07]
        integer :: k_s, i, n
        n = size(A, 1)
        n_inf = maxval(sum(abs(A), dim=2))
        k_s = max(0, int(ceiling(log(n_inf / 0.5_real64) / log(2.0_real64))))
        As = A / (2.0_real64**k_s)
        P = 0.0; Q = 0.0; Ap = 0.0; do i = 1, n; P(i, i) = 1.0; Q(i, i) = 1.0; Ap(i, i) = 1.0; end do
        do i = 1, 6
            Ap = matmul(Ap, As); P = P + Ap * c(i + 1)
            if (mod(i, 2) == 0) then; Q = Q + Ap * c(i + 1); else; Q = Q - Ap * c(i + 1); end if
        end do
        E = matmul(mat_i_dynamic(Q), P)
        do i = 1, k_s; E = matmul(E, E); end do
    end function exp_m_dynamic

    subroutine print_sorted_agents()
        ! 省略（必要に応じて実装）
    end subroutine print_sorted_agents

end program emergence_sim_fortran_dynamic

File: Part3/emergence_sim.py
import math
import sys
import random
import os

# ============================================================
# PKGF Intelligence Emergence Simulator (Part 3: Dynamic Dimensions)
# Directory: /private/test/PKGF_Intelligence_Emergence/Part3/
# Goal: Comparison of N=DIM={4, 8, 16}
# ============================================================

# --- MATRIX UTILS (N-DIM Generalized) ---

def identity_matrix(n): return [[1.0 if i == j else 0.0 for j in range(n)] for i in range(n)]
def zero_matrix(n): return [[0.0 for _ in range(n)] for _ in range(n)]
def mat_mul(A, B, n):
    C = zero_matrix(n)
    for i in range(n):
        for k in range(n):
            if abs(A[i][k]) > 1e-18:
                aik = A[i][k]
                for j in range(n): C[i][j] += aik * B[k][j]
    return C
def mat_inv(A, n):
    res = identity_matrix(n)
    m = [row[:] + res[i] for i, row in enumerate(A)]
    for i in range(n):
        pivot = m[i][i]
        if abs(pivot) < 1e-18:
            for k in range(i+1, n):
                if abs(m[k][i]) > abs(pivot):
                    m[i], m[k] = m[k], m[i]
                    pivot = m[i][i]; break
        if abs(pivot) < 1e-18: continue
        inv_p = 1.0 / pivot
        for j in range(2*n): m[i][j] *= inv_p
        for k in range(n):
            if i != k:
                f = m[k][i]
                for j in range(2*n): m[k][j] -= f * m[i][j]
    return [row[n:] for row in m]

def matrix_exp_pade(A, n):
    inf_norm = max(sum(abs(x) for x in row) for row in A)
    k = max(0, math.ceil(math.log2(inf_norm / 0.5))) if inf_norm > 0 else 0
    A_s = [[x / (2**k) for x in row] for row in A]
    c = [1.0, 0.5, 0.1, 0.01111111111111111, 0.0007575757575757576, 3.0303030303030303e-05, 5.050505050505051e-07]
    P, Q, A_p = identity_matrix(n), identity_matrix(n), identity_matrix(n)
    for i in range(1, 7):
        A_p = mat_mul(A_p, A_s, n)
        term = [[x * c[i] for x in row] for row in A_p]
        for r in range(n):
            for col in range(n):
                P[r][col] += term[r][col]
                if i % 2 == 0: Q[r][col] += term[r][col]
                else:          Q[r][col] -= term[r][col]
    res = mat_mul(mat_inv(Q, n), P, n)
    for _ in range(k): res = mat_mul(res, res, n)
    return res

# --- GEOMETRY ---

def get_metric_tensor(x, n):
    g = zero_matrix(n)
    # Contextual Warping: DIMを4セクターに分けて計量を変調
    # 最後の1/4セクターをContextとする
    ctx_idx = (n * 3) // 4
    s = [1.0 + 0.5 * math.tanh(x[i]) for i in range(ctx_idx, n)]
    avg_ctx = sum(s) / len(s) if s else 1.0
    for i in range(n):
        if i < ctx_idx: g[i][i] = avg_ctx
        else: g[i][i] = 1.0
    return g

def compute_christoffel_symbols(x, n):
    eps = 1e-5
    g_inv = mat_inv(get_metric_tensor(x, n), n)
    partials = []
    for k in range(n):
        xp, xm = x[:], x[:]; xp[k] += eps; xm[k] -= eps
        gp, gm = get_metric_tensor(xp, n), get_metric_tensor(xm, n)
        diff = [[(gp[i][j] - gm[i][j]) * 0.5 / eps for j in range(n)] for i in range(n)]
        partials.append(diff)
    Gamma = [[[0.0 for _ in range(n)] for _ in range(n)] for _ in range(n)]
    for k in range(n):
        for i in range(n):
            for j in range(n):
                val = 0.0
                for l in range(n):
                    val += g_inv[k][l] * (partials[i][j][l] + partials[j][i][l] - partials[l][i][j])
                Gamma[k][i][j] = 0.5 * val
    return Gamma

def parallel_transport_key(K, x, v_dt, dt, n):
    G = compute_christoffel_symbols(x, n); Omega = zero_matrix(n)
    for i in range(n):
        for j in range(n): Omega[i][j] = sum(G[i][k][j] * v_dt[k] for k in range(n))
    H = matrix_exp_pade([[val * dt for val in row] for row in Omega], n)
    return mat_mul(mat_mul(H, K, n), mat_inv(H, n), n)

# --- CORE AGENT ---

class Agent:
    def __init__(self, name, x_init, K_init, idx, n, dim):
        self.name, self.x, self.K = name, x_init, K_init
        self.idx, self.n, self.dim = idx, n, dim
        self.A_val, self.hunger, self.reward = 0.0, 0.0, 0.0
        self.C_vec = [1.0, 1.0, 0.0]
        self.mode = "Neutral"
        
        # 性格勾配
        ratio = idx / (n - 1)
        if ratio < 0.2: self.h_threshold, self.t_threshold, self.e_power = 0.2, 5.0, 0.01
        elif ratio < 0.5: self.h_threshold, self.t_threshold, self.e_power = 0.6, 1.2, 0.8
        else: self.h_threshold, self.t_threshold, self.e_power = 1.2, 0.6, 5.0

        # 親和性
        self.affinities = [math.cos(2.0 * math.pi * min(abs(idx-j), n-abs(idx-j)) / n) for j in range(n)]

    def decide(self, world):
        stress = 0.0
        for o in world.agents:
            if o == self: continue
            dist = math.sqrt(sum((self.x[k]-o.x[k])**2 for k in range(self.dim))) + 1e-9
            stress += -self.affinities[o.idx] / (dist + 0.5)
        
        ten = max(0, self.C_vec[2] + stress * 0.2)
        if self.hunger > self.h_threshold: self.mode = "Aggressive"
        elif ten > self.t_threshold: self.mode = "Submissive"
        else: self.mode = "Neutral"
        if self.reward > 0.1: self.hunger *= 0.5

    def get_v(self, world):
        dim = self.dim
        def field_omega(p):
            v_r = [(world.resource_pos[i] - p[i]) * 0.4 for i in range(dim)]
            g = get_metric_tensor(p, dim)
            return [sum(g[i][j] * v_r[j] for j in range(dim)) for i in range(dim)]
        
        # Co-differential Approximation
        omega_x = field_omega(self.x); eps = 1e-4; v_pkgf = [0.0]*dim
        for i in range(dim):
            pp = self.x[:]; pp[i] += eps; omega_p = field_omega(pp)
            for j in range(dim): v_pkgf[j] += (omega_p[j] - omega_x[j]) / eps # Simplification
        
        d_w = 0.3 + self.hunger * 3.0
        v_des = [(world.resource_pos[i] - self.x[i]) * d_w for i in range(dim)]
        
        v_eth = [0.0] * dim
        norm_factor = 2.0 / self.n
        for o in world.agents:
            if o == self: continue
            diff = [self.x[k] - o.x[k] for k in range(dim)]
            dist_sq = sum(d**2 for d in diff) + 1e-9
            aff = self.affinities[o.idx]
            e_s = self.e_power * (1.5 - aff)
            if self.mode == "Aggressive": e_s *= 0.01
            elif self.mode == "Submissive": e_s *= 2.0
            
            f = -e_s * norm_factor * (1.0 + self.C_vec[2]) / (dist_sq + 0.05)
            if aff > 0.5: f += 0.2 * aff / (math.sqrt(dist_sq) + 1.0)
            for k in range(dim): v_eth[k] += f * diff[k]
        
        eta = [random.uniform(-0.01, 0.01) for _ in range(dim)]
        return [v_pkgf[i] + v_des[i] + world.l_base * v_eth[i] + eta[i] for i in range(dim)]

    def update(self, v, dt, world):
        dim = self.dim
        d_p = math.sqrt(sum((self.x[i] - world.resource_pos[i])**2 for i in range(dim)))
        self.x = [self.x[i] + v[i] * dt for i in range(dim)]
        d_c = math.sqrt(sum((self.x[i] - world.resource_pos[i])**2 for i in range(dim)))
        self.K = parallel_transport_key(self.K, self.x, v, dt, dim)
        
        ex = 1.2 * max(0, (d_c - d_p)/dt) + 0.4 * math.sqrt(sum(vi**2 for vi in v))
        self.A_val = max(0, min(5.0, self.A_val + (-0.6 * self.A_val + ex) * dt))
        for i in range(dim): self.K[i][i] += random.uniform(-0.005, 0.005) * self.A_val
        
        tau = 0.35
        self.C_vec[2] += (-tau * self.C_vec[2] + math.tanh(self.A_val * 1.3)) * dt
        self.C_vec[2] = max(0, min(2.0, self.C_vec[2]))
        
        if d_c < 0.4: self.reward += 0.4 / (1.0 + self.A_val)
        else: self.hunger += 0.15 * dt

class World:
    def __init__(self, n):
        self.n, self.dim, self.agents = n, n, []
        self.resource_pos = [0.0]*n
        self.l_base, self.dt = 0.4, 0.1
        K_init = identity_matrix(n)
        for i in range(n):
            pos = [(random.random()-0.5)*2.0 for _ in range(n)]
            self.agents.append(Agent(f"A_{i:02d}", pos, [r[:] for r in K_init], i, n, n))
    def step(self):
        v_list = [a.get_v(self) for a in self.agents]
        for a in self.agents: a.decide(self)
        for a, v in zip(self.agents, v_list): a.update(v, self.dt, self)

# --- EXPERIMENTS ---

def run_experiment(n):
    print(f"\n--- Case: N={n}, DIM={n} ---")
    w = World(n)
    print(f"{'Step':<5} | {'Modes (A/N/S)':<15} | {'Avg_Rew':<10} | {'Avg_Ten':<10}")
    for s in range(201):
        modes = [a.mode[0] for a in w.agents]
        counts = f"{modes.count('A')}/{modes.count('N')}/{modes.count('S')}"
        rewards = [a.reward for a in w.agents]
        tensions = [a.C_vec[2] for a in w.agents]
        avg_r = sum(rewards)/n; avg_t = sum(tensions)/n
        if s % 40 == 0: print(f"{s:<5} | {counts:<15} | {avg_r:10.4f} | {avg_t:10.4f}")
        w.step()
    return avg_t

if __name__ == "__main__":
    log_dir = "/private/test/PKGF_Intelligence_Emergence/Part3/logs/python"
    os.makedirs(log_dir, exist_ok=True)
    sys.stdout = open(f"{log_dir}/phase_G_dimension_comparison.txt", "w")
    for n in [4, 8, 16]:
        run_experiment(n)
